agent: Simplify code using TryFrom trait

jodh-intel opened the pull request:
Simplified `parse_input()` drastically by implementing `TryFrom` for
`InputPayload` and `KeyProviderInput`.

Also refactored error handling and added lots of unit tests.

Signed-off-by: James O. D. Hunt <james.o.hunt@intel.com>

jodh-intel commented:
I think there is more opportunity for further refactoring (and additional tests), but this is a start :smile:

jialez0 reviewed (approved):
LGTM.

jodh-intel commented:
@jialez0 - Thanks for reviewing. I've just updated the branch further so that now `un_wrap_key()` calls `InputPayload::try_from()` rather than `parse_input()`, as that is more idiomatic and consistent with the rest of this branch.

jodh-intel commented:
@sameo, @fitzthum - Please could you tal?

fitzthum commented on src/grpc/protocol/keyprovider_structs.rs:
```diff
@@ -3,11 +3,24 @@
+pub const ERR_UNWRAP_PARAMS_NO_DC: &str = "keyunwrap parameters must include dc";
+pub const ERR_WRAP_PARAMS_EXPECT_EMPTY_EC: &str = "keywrap parameters should not include ec";
+pub const ERR_WRAP_PARAMS_EXPECT_EMPTY_OPTSDATA: &str =
+    "keywrap parameters should not include optsdata";
```
The Attestation Agent doesn't wrap keys, hence the comments in the original that say that the ec and opts should be null. I'm not sure if this is a requirement that we need to strictly enforce, however. A keywrap request without an encryption config is in fact not a valid keywrap request. Keywrap parameters should include and ec and optsdata, you just shouldn't ever be dealing with them in the AA. 

We have talked about trying to reuse some code between the AA and the sample KBS (which does wrap keys). These plans are tentative and not a huge priority, but would require reworking this to support (and require) keywrap requests with the necessary fields.

jodh-intel replied:
@fitzthum - ok. I'm not totally clear from your comment whether you'd like me to remove that additional checking?

fitzthum replied:
I think it's technically fine as it is. I guess we could skip the valid check on the keywrapparams. We don't really need some of the related functionality or tests but it doesn't hurt.

fitzthum commented on src/grpc/protocol/keyprovider_structs.rs:
```diff
@@ -17,34 +30,181 @@ pub struct KeyProviderInput {
+impl KeyProviderInput {
+    pub fn valid(&self) -> Result<()> {
+        match self.op.as_str() {
+            OP_KEY_WRAP => Err(anyhow!("{}: {:?}", ERR_UNSUPPORTED_OP, self.op)),
```
We might want this to be a more verbose message, suggesting that the user needs to use a different keyprovider (such as the sample kbs) to encrypt their images.

jodh-intel replied:
@fitzthum - Thanks. I've updated the message.

fitzthum reviewed (approved):
This looks great. 

I have one minor question regarding how we handle keywrap requests.

dubek commented on src/grpc/mod.rs:
```diff
@@ -99,60 +125,79 @@ impl KeyProviderService for KeyProvider {
+            .ok_or_else(|| anyhow!(ERR_ANNOTATION_EMPTY))?;
+
+        let vec_annotation = base64::decode(base64_annotation)?;
+
+        let jsonstring_annotation: &str = str::from_utf8(&vec_annotation)?;
```
shouldn't we check at this point that the `jsonstring_annotation` is valid JSON?

jodh-intel replied:
I believe that string should be just plain text at this point. I think that ariable name is erroneous, I've renamed it to `annotation` to avoid the confusion, but it would be good to get confirmation from @jialez0.

jodh-intel commented on src/grpc/protocol/keyprovider_structs.rs:
```diff
@@ -17,34 +30,181 @@ pub struct KeyProviderInput {
-#[derive(Serialize, Deserialize, Debug)]
+impl KeyUnwrapParams {
+    pub fn valid(&self) -> Result<()> {
+        // "empty" means "not defined" or "not set"
+        let dc_empty = if let Some(dc) = self.dc.clone() {
```
It's because `Dc` doesn't implement `Copy` - try removing the `clone()` and the compiler will bleat at you :smile:

jodh-intel commented:
Branch updated with the following changes:

- global: Added new `ERR_ANNOTATION_NOT_BASE64` and `ERR_KBC_KBS_NOT_BASE64` errors.
- `InputPayload`: Simplified `try_from()`.
- `Dc`: Added in check for valid URI (and new unit tests).
- tests: Improved failure messages.
- Fixed a potential crasher bug if the Dc parameters vector was empty (I've added a unit test).

I've also found a problem: as noted above the code now checks the KBS URI to make sure it looks valid. I've added a few unit tests but then I found https://github.com/confidential-containers/attestation-agent/tree/main/src/kbc_modules/eaa_kbc#usage which includes:

> Set KBC_NAME::KBS_URI pair in the following format:
> ```
> eaa_kbc::<IP>:<PORT>
> ```
>
> e.g: EAA KBS address is 127.0.0.1:1122 :
>
> ``` 
> eaa_kbc::127.0.0.1:1122
> ```

The problem is `eaa_kbc` is not a valid URI scheme since schemes can only use alphabetics, `+`, `-` and `.` (see https://datatracker.ietf.org/doc/html/rfc3986#section-3.1).

Can I just go ahead and "fix" that document so the scheme becomes `eaa-kbc`? If so, I'll also add unit tests for `eaa_kbc` and `eaa-kbc`.

Jakob-Naucke commented:
> The problem is `eaa_kbc` is not a valid URI scheme since schemes can only use alphabetics, `+`, `-` and `.` (see https://datatracker.ietf.org/doc/html/rfc3986#section-3.1).
> 
> Can I just go ahead and "fix" that document so the scheme becomes `eaa-kbc`? If so, I'll also add unit tests for `eaa_kbc` and `eaa-kbc`.

Hmm. That would apply to all KBCs implemented thus far. We used underscores because Rust modules generally use snake casing and didn't really consider URL validity. We could, but users should be very aware of this change because their configurations will need updating. @fitzthum WDYT?

jodh-intel commented:
It would be a real shame if we didn't comply with the URI spec though :cry:

fitzthum commented:
We're expecting`KBC_NAME::KBS_URI`. I think only the second half is supposed to be the URI. `KBC_NAME` isn't the scheme so I think its format may be fine. That said, we aren't actually providing a URI in `KBS_URI` in that example because it has no scheme. 

Originally the idea with a URI was to differentiate KBCs that point to external resources like KBS and KBCs that point to internal resources like files. We ended up hardcoding the path for our two file-based KBCs so this feature wasn't given much scrutiny. Based on my initial understanding, we might want to update the `KBC_URI` to have a scheme.

It's also slightly weird that the `KBS_URI` is pointing to localhost in that example. That would seem to suggest that the KBS is running inside the enclave, which is odd.

jodh-intel commented:
Hi @fitzthum - thanks for explaining the context here. We still have a problem though because a URI requires a scheme, so `127.0.0.1:1122` is "invalid" as there is no scheme prefix, hence some questions:

- Do we / can we assume a default scheme to allow the URI to be validated?
- `<IP>`: Do we mean IPv4 *and* IPv6?

I'd like to be able to validate the URI but without some sort of scheme we can't rely on standard facilities.

/cc @sameo.

jodh-intel commented:
/cc @jialez0

jialez0 commented:
@jodh-intel I'm sorry that we didn't consider the URI format specification before. As you said, if we need to verify the URI here with the standard facilities, maybe we need to add an appropriate scheme prefix to socket address such as `<IPv4/v6>:<port>`, but I don't know the URI format specification, so I'm not sure what kind of scheme prefix it should be.

dubek commented:
I think this is the address for the KBS grpc endpoint. grpc is implemented over http2 if I remember correctly (but please check). If this is true, then an `https://` scheme should work:   `https://IPv4:PORT/` (not sure how ipv6 works in a URI).

jodh-intel commented:
Branch rebased to resolve the conflict introduced by #31.

Thanks for the comments, but I'm not convinced about the URL validation. It sounds like the topic needs further discussion to me, so I've removed that part of the branch. I'm happy to add it at a later date.

jodh-intel commented:
Fixed clippy test code nits.

jodh-intel commented:
All green ;)
